In [1]:
# A look into FPI activity in equity from NSDL
# Data is published with an 8 month delay.

In [1]:
from pathlib import Path

import pandas as pd

In [74]:
%load_ext jupyter_black
pd.options.display.float_format = "{:.0f}".format

### Explore a single month of data

In [16]:
DATA_PATH = Path() / "nsdl_data" / "2025"

In [25]:
dat1 = pd.read_excel(DATA_PATH / "jan-2025.xlsx", engine="openpyxl")

In [26]:
dat1.head()

RFDE_CUST_REG_NUM RFDE_RPT_DT   RFDE_TXN_ID             FII         SUB_ACC  \
0        IN/CUS/025  2025-01-01           184             NaN   S868343869412   
1        IN/CUS/022  2025-01-01       1085893   F750013435337  S1200737910842   
2        IN/CUS/022  2025-01-01       1085894   F586108693778   S421851293887   
3        IN/CUS/023  2025-01-01  000063615929  F1068518164010  S1048660787178   
4        IN/CUS/023  2025-01-01  000063615927   F874962995022   S502610051852   

           BRKER                      SCRIP_NAME          ISIN    TR_DATE  \
0  B770386235885                        NMDC LTD  INE584A01023 2025-01-01   
1  B770386235885     Banco Products (India) Ltd.  INE213C01025 2025-01-01   
2  B770386235885     Banco Products (India) Ltd.  INE213C01025 2025-01-01   
3  B770386235885  BANCO PRODUCTS (INDIA) LIMITED  INE213C01025 2025-01-01   
4  B770386235885  BANCO PRODUCTS (INDIA) LIMITED  INE213C01025 2025-01-01   

   TR_TYPE(*)  RFDE_SE_REG_NUM  RFDE_STLD_CODE  RATE  QUANTITY  VALUE (in Rs)  \
0           7               99  REG_DL_STLD_04   0.0    7298.0            0.0   
1           7               99  REG_DL_STLD_04   0.0     930.0            0.0   
2           7               99  REG_DL_STLD_04   0.0   13326.0            0.0   
3           7               99  REG_DL_STLD_04   0.0     587.0            0.0   
4           7               99  REG_DL_STLD_04   0.0  111457.0            0.0   

   RFDE_INSTR_TYPE RFDE_REASON_DLAY  RFDE_RPT_TYPE RFDE_AMDMNT_REASON  
0  REG_DL_INSTR_EQ    REG_DL_DLAY_P  DL_RPT_TYPE_N   DL_AMDMNT_DEL_00  
1  REG_DL_INSTR_EQ    REG_DL_DLAY_P  DL_RPT_TYPE_N   DL_AMDMNT_DEL_00  
2  REG_DL_INSTR_EQ    REG_DL_DLAY_P  DL_RPT_TYPE_N   DL_AMDMNT_DEL_00  
3  REG_DL_INSTR_EQ    REG_DL_DLAY_P  DL_RPT_TYPE_N   DL_AMDMNT_DEL_00  
4  REG_DL_INSTR_EQ    REG_DL_DLAY_P  DL_RPT_TYPE_N   DL_AMDMNT_DEL_00

In [65]:
dat1.rename(columns={"TR_TYPE(*)": "TR_TYPE", "VALUE (in Rs)": "VALUE"}, inplace=True)

In [66]:
dat1.iloc[0].transpose()

RFDE_CUST_REG_NUM              IN/CUS/025
RFDE_RPT_DT           2025-01-01 00:00:00
RFDE_TXN_ID                           184
FII                                   NaN
SUB_ACC                     S868343869412
BRKER                       B770386235885
SCRIP_NAME                       NMDC LTD
ISIN                         INE584A01023
TR_DATE               2025-01-01 00:00:00
TR_TYPE                                 7
RFDE_SE_REG_NUM                        99
RFDE_STLD_CODE             REG_DL_STLD_04
RATE                                  0.0
QUANTITY                           7298.0
VALUE                                 0.0
RFDE_INSTR_TYPE           REG_DL_INSTR_EQ
RFDE_REASON_DLAY            REG_DL_DLAY_P
RFDE_RPT_TYPE               DL_RPT_TYPE_N
RFDE_AMDMNT_REASON       DL_AMDMNT_DEL_00
Name: 0, dtype: object

In [34]:
dat1.TR_TYPE.value_counts(ascending=False)

TR_TYPE
4     236836
1     196471
7       2772
15      1806
2        810
3         34
5         16
8         13
12        13
16        10
14         9
13         8
10         5
11         1
Name: count, dtype: int64

### Transaction type (TR_TYPE) - Frequent (top 2)
- Sale in secondary market 	04
- Purchase in secondary market 	01

In [72]:
fpi = dat1[(dat1.TR_TYPE == 1) | (dat1.TR_TYPE == 4)]

In [74]:
fpi.loc[:, "TR_DATE"] = pd.to_datetime(fpi["TR_DATE"])

In [61]:
fpi[fpi.ISIN == "INE982J01020"][:3]

RFDE_CUST_REG_NUM RFDE_RPT_DT   RFDE_TXN_ID             FII  \
691         IN/CUS/023  2025-01-02  0O00004G5YIS  F1068518164010   
742         IN/CUS/023  2025-01-02  0O00004G5W85  F1068518164010   
1051        IN/CUS/023  2025-01-02  0O00004G5WF3  F1068518164010   

             SUB_ACC           BRKER                     SCRIP_NAME  \
691   S1048660787178  B1083872365222  ONE 97 COMMUNICATIONS LIMITED   
742   S1048660787178  B1083872365222  ONE 97 COMMUNICATIONS LIMITED   
1051  S1048660787178  B1083872365222  ONE 97 COMMUNICATIONS LIMITED   

              ISIN    TR_DATE  TR_TYPE  RFDE_SE_REG_NUM  RFDE_STLD_CODE  \
691   INE982J01020 2025-01-01        1               23  REG_DL_STLD_01   
742   INE982J01020 2025-01-01        4                1  REG_DL_STLD_01   
1051  INE982J01020 2025-01-01        4               23  REG_DL_STLD_01   

          RATE  QUANTITY  VALUE (in Rs)  RFDE_INSTR_TYPE RFDE_REASON_DLAY  \
691   989.3106   33604.0    33244793.40  REG_DL_INSTR_EQ    REG_DL_DLAY_P   
742   986.2558      43.0       42409.00  REG_DL_INSTR_EQ    REG_DL_DLAY_P   
1051  988.1672    2843.0     2809359.35  REG_DL_INSTR_EQ    REG_DL_DLAY_P   

      RFDE_RPT_TYPE RFDE_AMDMNT_REASON  
691   DL_RPT_TYPE_N   DL_AMDMNT_DEL_00  
742   DL_RPT_TYPE_N   DL_AMDMNT_DEL_00  
1051  DL_RPT_TYPE_N   DL_AMDMNT_DEL_00

## ISIN maps to NSE
ISIN can be used to link to NSE if needed to obtain the stock information.
https://www.nseindia.com/get-quotes/equity?symbol=FIVESTAR  

There is `[FII, SUB_ACC, BRKER]` FII, Subaccount, and a broker associated with each data point.  
There is a potential to map this data as well.


## Initial analysis
To begin with, let's aim to summarize the purchase and sale by:
1. scrip
2. fpi
3. month/year

Build a quick shiny app for the above to be served from AWS.

In [91]:
fpi.to_csv("jan_buy_sell.csv")

In [90]:
fpi.groupby("TR_TYPE")[["VALUE"]].sum() / 10000000

VALUE
TR_TYPE          
1       262487.82
4       338538.38

## TODO - PLOTS
1. FPI monthly buy and sell value
2. FPI by country buy and sell value (monthly)
3. Scrip level diving of purchase and sale.
4. FPI level buy and sale or net position
5. Size of FPI and net position in scrips.

In [9]:
fpi = pd.read_csv("nsdl_data/jan_buy_sell_2025.csv", index_col=0)

In [10]:
cols_keep = ["SCRIP_NAME", "ISIN", "TR_DATE", "TR_TYPE", "RATE", "QUANTITY", "VALUE"]

In [11]:
fpi = fpi[cols_keep]

In [34]:
fpi[fpi.TR_TYPE == 4].groupby(["SCRIP_NAME", "ISIN"], as_index=True).agg(
    # {
    {"VALUE": "sum", "RATE": ["min", "max"]}
    TOTAL_VALUE = pd.NamedAgg(column="VALUE", aggfunc="sum"),
    LOWEST_PRICE =pd.NamedAgg(column="RATE", aggfunc="min"),
    # WEGHTED_RATE =     
# }
).sort_values(by=("VALUE", "sum"), ascending=False).head()

VALUE RATE     
                                                              sum  min  max
SCRIP_NAME                               ISIN                              
HDFC BANK LTD INR 1                      INE040A01034 77634791262 1626 1797
AXIS BANK LIMITED INR 2                  INE238A01034 52442149927  939 1089
HDFC BANK LIMITED EQ FV01                INE040A01034 49827688658 1626 1800
ICICI BANK LIMITED INR 2                 INE090A01021 44116598042 1188 1290
AXIS BANK LIMITED EQ FV 2                INE238A01034 39146858863  937 1089

## alculating the average monthly volume of FPIs

In [1]:
import pyarrow.parquet as pq

In [9]:
df1 = pq.read_table("fpi_data/fpi_2024_2025.parquet")

In [10]:
dfp = df1.to_pandas()

In [11]:
dfp.shape

(5908077, 13)

In [12]:
dfp = dfp[(dfp.TR_TYPE == 4) | (dfp.TR_TYPE == 1)]

In [13]:
dfp.head()

SCRIP_NAME          ISIN     TR_DATE  TR_TYPE       RATE  \
11  GUJARAT INDUSTRIES P  INE162A01010  2024-02-01        4   222.7558   
12  ADANI TOTAL GAS LIMI  INE399L01023  2024-02-01        1  1009.0338   
13  RESPONSIVE INDUSTRIE  INE688D01026  2024-02-01        1   299.7814   
14  RESPONSIVE INDUSTRIE  INE688D01026  2024-02-01        1   299.3705   
15  RESPONSIVE INDUSTRIE  INE688D01026  2024-02-01        4   300.9500   

    QUANTITY         VALUE month  year  mutual_fund  value_crores  \
11    5000.0  1.113779e+06   feb  2024            0      0.111378   
12  950000.0  9.585821e+08   feb  2024            0     95.858211   
13    4000.0  1.199126e+06   feb  2024            0      0.119913   
14   22408.0  6.708294e+06   feb  2024            0      0.670829   
15    4000.0  1.203800e+06   feb  2024            0      0.120380   

                                company_name instrument_type  
11  Gujarat Industries Power Company Limited          Equity  
12                   Adani Total Gas Limited          Equity  
13             Responsive Industries Limited          Equity  
14             Responsive Industries Limited          Equity  
15             Responsive Industries Limited          Equity

In [14]:
dfp.shape

(5841568, 13)

In [ ]:
dfp1_4 = dfp[(dfp.TR_TYPE == 4) | (dfp.TR_TYPE == 1)]

In [9]:
(
    dfp1_4.groupby(["month", "year", "TR_TYPE"])["QUANTITY"].sum() / 10000000
).reset_index()["QUANTITY"].mean()

np.float64(716.6711378341)

In [15]:
(
    dfp1_4[dfp1_4["year"] == 2024].pivot_table(
        index=["month", "year"],
        columns=["TR_TYPE"],
        values=["QUANTITY"],
        aggfunc="sum",
    )
    / 10000000
)

QUANTITY             
TR_TYPE              1            4
month year                         
apr   2024  866.226858   798.108041
aug   2024  987.912810  1035.010921
dec   2024  505.468094   554.735808
feb   2024  844.195422   815.491761
jan   2024  664.295956   703.615247
jul   2024  712.413343   693.121568
jun   2024  921.244076   798.978711
mar   2024  695.748968   596.767983
may   2024  859.784389   877.654533
nov   2024  478.867108   679.119795
oct   2024  558.497528   793.764521
sep   2024  823.008980   758.259924

volume: 743,13,86,277 of shares bought by FPIs on average each month.

In [19]:
(
    dfp1_4.pivot_table(
        index=["year"],# "month"],
        columns=["TR_TYPE"],
        values=["QUANTITY"],
        aggfunc="sum",
    )
    / 10000000 #crores
)

QUANTITY             
TR_TYPE            1            4
year                             
2024     8917.663532  9104.628814
2025     1650.809584  1827.032205

In [11]:
(
    dfp1_4[dfp1_4["year"] == 2024].pivot_table(
        index=["year"],
        columns=["TR_TYPE"],
        values=["value_crores"],
        aggfunc="sum",
    )
    # / 10000000
).reset_index()["value_crores"].mean()

TR_TYPE
1    4.178718e+06
4    4.296630e+06
dtype: float64

In [34]:
dfp.columns

Index(['SCRIP_NAME', 'ISIN', 'TR_DATE', 'TR_TYPE', 'RATE', 'QUANTITY', 'VALUE',
       'month', 'year', 'mutual_fund', 'value_crores', 'company_name',
       'instrument_type'],
      dtype='object')

## NSE total traded quantity calculations

https://www.nseindia.com/market-data/business-growth-cm-segment

In [40]:
total_qty_traded = 10427995 * 100000

In [43]:
fpi_bough_sold_shares = 180222923458

In [44]:
fpi_bough_sold_shares / total_qty_traded

0.17282605472864151

In [46]:
fpi_bough_sold_shares / 10000000

18022.2923458

## additional analysis using FII id data

In [3]:
import pandas as pd
import pyarrow.parquet as pq

In [6]:
dt1 = pq.read_table("fpi_data/fpi_id_2024_2025_mgd_jan2026.parquet").to_pandas()

In [7]:
dt1 = dt1[(dt1.TR_TYPE == 4) | (dt1.TR_TYPE == 1)]

In [8]:
dt1.FII.unique()

array(['F361420591500', None, 'F577817169357', ..., 'F960037035661',
       'F1052569156393', 'F598426607826'], shape=(33014,), dtype=object)

In [5]:
dt1[dt1.mutual_fund==0].tail(20)

SCRIP_NAME          ISIN            FII  \
5908006  ADITYA BIRLA FASHION N RETAIL EQ 10  INE647O01011           None   
5908007     INDIAN OVERSEAS BANK ORD EQ FV10  INE565A01014           None   
5908008    SYSTEMATIX CORP LTD SPLIT EQ FV10  INE356B01024           None   
5908010              DESCO INFRATECH EQ FV10  INE0TGG01014           None   
5908011               BANK OF BARODA EQ FV 2  INE028A01039           None   
5908013          SAMMAAN CAPITAL LTD EQ FV 2  INE148I01020           None   
5908014             IOB EQ LSTNG TRD APP AWT  IN8565A01013           None   
5908015   SBI CARDS N PAYMENT SR LTD EQ FV10  INE018E01016           None   
5908017             MAHINDRA AND MAH EQ FV 5  INE101A01026  F514548180004   
5908018   PUNJAB AND SIND TRD APRL AWAT FV10  IN8608A01011           None   
5908019     INDIAN OVERSEAS BANK ORD EQ FV10  INE565A01014           None   
5908022             IOB EQ LSTNG TRD APP AWT  IN8565A01013           None   
5908023          MARUTI SUZUKI INDIA EQ FV 5  INE585B01010  F514548180004   
5908024             MAHINDRA AND MAH EQ FV 5  INE101A01026           None   
5908027  CENTRAL BANK OF INDIA APRL AWT FV10  IN8483A01019           None   
5908028            PRAXIS HOME RETAIL EQ FV5  INE546Y01022           None   
5908033          MARUTI SUZUKI INDIA EQ FV 5  INE585B01010  F514548180004   
5908035            RACL GEARTECH LTD EQ FV10  INE704B01017           None   
5908036    SYSTEMATIX CORP LTD SPLIT EQ FV10  INE356B01024           None   
5908037            ELEGANZ INTERIORS EQ FV10  INE0R9101015           None   

                SUB_ACC     TR_DATE  TR_TYPE        RATE   QUANTITY  \
5908006   S535219383756  2025-03-28        1    256.3109    10000.0   
5908007  S1024003831523  2025-03-28        1      0.0000  1725412.0   
5908008            None  2025-03-28        1    140.0000    10000.0   
5908010  S1078882893452  2025-03-28        1    150.0000   176000.0   
5908011            None  2025-03-27        1      0.0000    32175.0   
5908013            None  2025-03-28        1    107.5168    17411.0   
5908014            None  2025-03-28        4      0.0000  2464875.0   
5908015            None  2025-03-28        1    882.8556      167.0   
5908017   S565005449005  2025-03-28        4   2657.8330     4427.0   
5908018  S1078882893452  2025-03-28        1     38.3700  1824341.0   
5908019            None  2025-03-28        4     39.1976   688400.0   
5908022  S1024003831523  2025-03-28        4      0.0000  1725412.0   
5908023   S565005449005  2025-03-28        1  11594.3196     1025.0   
5908024            None  2025-03-28        4   2650.1830     1100.0   
5908027            None  2025-03-29        1     40.4900  1961470.0   
5908028            None  2025-03-28        1      8.9000  1340000.0   
5908033   S565005449005  2025-03-28        1  11594.3196      284.0   
5908035   S356855650596  2025-03-28        1    748.6000      371.0   
5908036            None  2025-03-28        4    141.1957     1833.0   
5908037            None  2025-03-28        4    123.0333     3000.0   

               VALUE month  year  mutual_fund  value_crores  \
5908006   2563109.00   mar  2025            0      0.256311   
5908007         0.00   mar  2025            0      0.000000   
5908008   1400000.00   mar  2025            0      0.140000   
5908010  26400000.00   mar  2025            0      2.640000   
5908011         0.00   mar  2025            0      0.000000   
5908013   1871975.00   mar  2025            0      0.187197   
5908014         0.00   mar  2025            0      0.000000   
5908015    147436.89   mar  2025            0      0.014744   
5908017  11766226.69   mar  2025            0      1.176623   
5908018  69999964.17   mar  2025            0      6.999996   
5908019  26983627.84   mar  2025            0      2.698363   
5908022         0.00   mar  2025            0      0.000000   
5908023  11884177.59   mar  2025            0      1.188418   
5908024   2915201.30   mar  2025          

SCRIP_NAME          ISIN            FII  \
97669    INTELLIVATE CAPITAL VENTURES EQ 2  IN8512D01019  F952611593286   
97678     INTELLIVATE CAPITAL VENT EQ FV 1  INE512D13015  F952611593286   
223017    INDIABULLS HOUS FIN RGHTS EQ FV2  INE148I20012  F952611593286   
244875          INTELLIVATE CAPITAL EQ FV1  INE512D01028  F952611593286   
244877   INTELLIVATE CAPITAL VENTURES EQ 2  IN8512D01019  F952611593286   
...                                    ...           ...            ...   
5907987              AXIS BANK LTD EQ FV 2  INE238A01034           None   
5908007   INDIAN OVERSEAS BANK ORD EQ FV10  INE565A01014           None   
5908011             BANK OF BARODA EQ FV 2  INE028A01039           None   
5908014           IOB EQ LSTNG TRD APP AWT  IN8565A01013           None   
5908022           IOB EQ LSTNG TRD APP AWT  IN8565A01013           None   

                SUB_ACC     TR_DATE  TR_TYPE  RATE   QUANTITY  VALUE month  \
97669     S647155222887  2024-02-08        1   0.0    50225.0    0.0   feb   
97678     S647155222887  2024-02-08        4   0.0    50225.0    0.0   feb   
223017    S647155222887  2024-02-20        4   0.0     3000.0    0.0   feb   
244875    S647155222887  2024-02-21        1   0.0    50225.0    0.0   feb   
244877    S647155222887  2024-02-21        4   0.0    50225.0    0.0   feb   
...                 ...         ...      ...   ...        ...    ...   ...   
5907987            None  2025-03-27        1   0.0     3750.0    0.0   mar   
5908007  S1024003831523  2025-03-28        1   0.0  1725412.0    0.0   mar   
5908011            None  2025-03-27        1   0.0    32175.0    0.0   mar   
5908014            None  2025-03-28        4   0.0  2464875.0    0.0   mar   
5908022  S1024003831523  2025-03-28        4   0.0  1725412.0    0.0   mar   

         year  mutual_fund  value_crores          company_name instrument_type  
97669    2024            0           0.0                  None            None  
97678    2024            0           0.0                  None            None  
223017   2024            0           0.0                  None            None  
244875   2024            0           0.0                  None            None  
244877   2024            0           0.0                  None            None  
...       ...          ...           ...                   ...             ...  
5907987  2025            0           0.0     Axis Bank Limited          Equity  
5908007  2025            0           0.0  Indian Overseas Bank          Equity  
5908011  2025            0           0.0        Bank Of Baroda          Equity  
5908014  2025            0           0.0                  None            None  
5908022  2025            0           0.0                  None            None  

[1575 rows x 15 columns]

> There are 1575 rows with 0 values. data reporting issues. In these cases, it is either the price or quantity is zero.

In [11]:
dt1 = dt1[dt1.VALUE!=0]

> There are many blanks FII IDs (there is a sub account number). I'd like to these records even if it will severely underestimate the count of FII trading on a given stock of interest.

In [75]:
# this is the value of records without a FII
dt1[dt1["FII"].isna()]["VALUE"].sum()

np.float64(20118047148118.8)

In [76]:
# this is the value of records without a SUB_ACC
dt1[dt1["SUB_ACC"].isna()]["VALUE"].sum()

np.float64(8876022630149.97)

In [77]:
# this is % of total value accounted by records without SUBACC ID
(dt1[dt1["SUB_ACC"].isna()]["VALUE"].sum()) / (dt1["VALUE"].sum())

np.float64(0.08624919727745052)

In [78]:
# this is % of total value accounted by records without FII ID
(dt1[dt1["FII"].isna()]["VALUE"].sum()) / (dt1["VALUE"].sum())

np.float64(0.19548907090673243)

In [79]:
# this is % of total value accounted by records with FII
(dt1[(dt1["FII"].isna()) & (dt1["SUB_ACC"].isna())]["VALUE"].sum()) / (
    dt1["VALUE"].sum()
)

np.float64(0.08111166679279104)

In [80]:
# which stocks are most impacted by missing values in FII and Sub_ACC
dt1[(dt1["FII"].isna()) & (dt1["SUB_ACC"].isna()) & (dt1["mutual_fund"] == 0)].groupby(
    "ISIN"
).count()

SCRIP_NAME  FII  SUB_ACC  TR_DATE  TR_TYPE  RATE  QUANTITY  \
ISIN                                                                       
IN0020210244           1    0        0        1        1     1         1   
IN800LO01018           1    0        0        1        1     1         1   
IN8024L01018           1    0        0        1        1     1         1   
IN80F3301011           1    0        0        1        1     1         1   
IN8160A01021           2    0        0        2        2     2         2   
...                  ...  ...      ...      ...      ...   ...       ...   
INE998U01015          12    0        0       12       12    12        12   
INE999A01015         146    0        0      146      146   146       146   
INE999A01023         436    0        0      436      436   436       436   
INE999B01013           4    0        0        4        4     4         4   
INE999K01014         628    0        0      628      628   628       628   

              VALUE  month  year  mutual_fund  value_crores  company_name  \
ISIN                                                                        
IN0020210244      1      1     1            1             1             1   
IN800LO01018      1      1     1            1             1             0   
IN8024L01018      1      1     1            1             1             0   
IN80F3301011      1      1     1            1             1             0   
IN8160A01021      2      2     2            2             2             0   
...             ...    ...   ...          ...           ...           ...   
INE998U01015     12     12    12           12            12            12   
INE999A01015    146    146   146          146           146             0   
INE999A01023    436    436   436          436           436           436   
INE999B01013      4      4     4            4             4             4   
INE999K01014    628    628   628          628           628           628   

              instrument_type  
ISIN                           
IN0020210244                1  
IN800LO01018                0  
IN8024L01018                0  
IN80F3301011                0  
IN8160A01021                0  
...                       ...  
INE998U01015               12  
INE999A01015                0  
INE999A01023              436  
INE999B01013                4  
INE999K01014              628  

[2433 rows x 14 columns]

In [82]:
# check the distribution of the 3 digit in the ISIN - indicates govt, state, DVR

dt1[dt1.ISIN.str.contains("^IN[0-9]", regex=True)]["ISIN"].apply(
    lambda x: x[2]
).value_counts()

ISIN
9    7530
8      93
0       1
Name: count, dtype: int64

In [41]:
dt1[~dt1['FII'].isna()].groupby('year')['FII'].unique().apply(lambda x: len(x))

year
2024    26571
2025     6442
Name: FII, dtype: int64

## what more?

what can I do here? the fii ids can be used. Perhaps build a model straight out and not waste any more time on this summarization.

predict FII activity? 
is there clustering of FII activity in stocks? if one FII sells is it followed more selling by FIIs. There is possibly a mix of stories here .

In [86]:
months_order = [
    "jan",
    "feb",
    "mar",
    "apr",
    "may",
    "jun",
    "jul",
    "aug",
    "sep",
    "oct",
    "nov",
    "dec",
]

In [87]:
dt1["month"] = pd.Categorical(dt1["month"], categories=months_order, ordered=True)

In [97]:
dt1[dt1.ISIN == 'INE585B01010'].pivot_table(
        index=[ "year"],
        columns=["TR_TYPE"],
        values=["value_crores"],
        aggfunc="sum",
    observed=False
    )

value_crores              
TR_TYPE             1             4
year                               
2024     32305.922242  50009.234391
2025      5866.459041   7756.831780

In [94]:
dt1[dt1.ISIN == 'INF205K011T7'].pivot_table(
        index=[ "year"],
        columns=["TR_TYPE"],
        values=["value_crores"],
        aggfunc="sum",
    observed=False
    ).diff(axis=1)["value_crores"] #.sort_values() #iloc[0]  #.values[:,1]

TR_TYPE   1
year       
2025    NaN

### number of FIIs in a stock

In [9]:
def fii_count_stock(ISIN_NO):
    dt_isin = dt1[dt1["ISIN"] == ISIN_NO]
    dt_isin = dt_isin.drop_duplicates(["month", "year", "FII"])
    dt_isin = (
        (
            dt_isin.groupby(["month", "year", "TR_TYPE"], observed=False)["FII"]
            .count()
            .reset_index()
        )
        .pivot_table(
            columns="TR_TYPE", index=["month", "year"], values="FII", observed=False
        )
        .reset_index()
        .sort_values(by=["year", "month"])
    )
    return dt_isin

In [11]:
dt = fii_count_stock("INE238A01034")

In [12]:
dt.rename(columns={1:"Buy", 4:"Sell"})

TR_TYPE month  year    Buy   Sell
0         apr  2024  215.0  289.0
1         aug  2024  201.0  273.0
2         dec  2024  176.0  248.0
3         feb  2024  241.0  273.0
5         jan  2024  233.0  265.0
7         jul  2024  241.0  294.0
8         jun  2024  271.0  214.0
9         mar  2024  194.0  289.0
11        may  2024  240.0  259.0
12        nov  2024  174.0  265.0
13        oct  2024  202.0  300.0
14        sep  2024  182.0  258.0
4         feb  2025  197.0  229.0
6         jan  2025  167.0  303.0
10        mar  2025  234.0  185.0